### PR TITLE
Fix UnlockOnActivity for concurrent requests

### PR DESCRIPTION
**Pull request by archimed7592**
Hello!

Here is a fix for scenario when database is locked and multiple requests are being sent to the agent.
For example, if you have a git repo with multiple ssh remotes, and you do a "fetch all remotes" in your IDE.

It is easily reproduced with multiple ssh-add -l commands thrown at the agent.

Behavior before fix:
* One password prompt is shown
* Only first "ssh-add -l" request is granted
* Subsequent requests receive an empty list of keys

Behavior after fix:
* One password prompt is shown
* All of the request are granted with a correct list of keys

### CODE REVIEW

File: KeeAgent/KeeAgentExt.cs [hunk 1]
```diff
@@ -815,35 +815,29 @@ private void PageantAgent_KeyRemoved(object sender, SshKeyEventArgs e)
       RemoveKey(e.Key);
     }
 
+    private object unlockOnActivitySyncRoot = new object(); // used in PageantAgent_MessageReceived as a sync root
```
Comment (dlech) [inline KeeAgent/KeeAgentExt.cs:818]: It looks like this can be `readonly`.

File: KeeAgent/KeeAgentExt.cs [hunk 2]
```diff
@@ -815,35 +815,29 @@ private void PageantAgent_KeyRemoved(object sender, SshKeyEventArgs e)
...
-        {
-          mainWindow.Invoke(
-            (MethodInvoker)delegate()
-          {
-            // don't do anything - we are just seeing if the thread is blocked
-          });
-        });
-      thread.Name = "Check";
-      thread.Start();
-      // only try to unlock databases if main thread is not blocked
```
Comment (dlech) [inline KeeAgent/KeeAgentExt.cs:833]: Does removing this extra thread check cause a regression of https://github.com/dlech/KeeAgent/issues/19?

**Review by dlech — COMMENTED** [review]
Thanks for the PR. There are some nice improvements here. I have one concern about causing a regression though. When the IOProtocol extension is installed, it can block the main thread waiting for the agent to respond to a request when it tries to perform an sftp/scp action.

File: KeeAgent/KeeAgentExt.cs [hunk 3]
```diff
@@ -815,35 +815,29 @@ private void PageantAgent_KeyRemoved(object sender, SshKeyEventArgs e)
...
-      if (thread.Join(1000)) {
-        mainWindow.Invoke((MethodInvoker)delegate()
-        {
-          if (Options.UnlockOnActivity) {
-            foreach (var document in mainWindow.DocumentManager.Documents) {
+      lock (unlockOnActivitySyncRoot) {
+        // ↑ protection from concurrent requests, for example "git fetch" with multiple ssh remotes
+
+        mainWindow.Invoke((MethodInvoker)delegate () {
```
Comment (dlech) [inline KeeAgent/KeeAgentExt.cs:834]: We can change this to an `=>` expression while we are touching this.

File: KeeAgent/KeeAgentExt.cs (hunk 1 above)
Comment (archimed7592) [reply KeeAgent/KeeAgentExt.cs]: @dlech , fixed

File: KeeAgent/KeeAgentExt.cs (hunk 3 above)
Comment (archimed7592) [reply KeeAgent/KeeAgentExt.cs]: @dlech , fixed

File: KeeAgent/KeeAgentExt.cs (hunk 2 above)
Comment (archimed7592) [reply KeeAgent/KeeAgentExt.cs:833]: @dlech , my bad.
I thought this check had something to do with the problem I was dealing with and I wasn't smart enough to have a look at git blame to find out why this check was added in the first place.
Fixed.

**Comment by dlech** [comment]
Thanks!

**Pull request merged**